### PR TITLE
Only involve tensor device in CUDA -> CPU copy, not current device.

### LINKED ISSUE #10832: CUDA context created on gpu0 when only gpu1 is used
```py
x = torch.randn(1).cuda(1)
y = torch.randn(1).cuda(1)
# here only gpu 1 has memory used
torch.cat([x,y],0)
# here gpu 0 has memory used as well!
```

```py
torch.randn(3, device='cuda:1')
# uses both GPU 0 and 1 memory!
```

### PR DESCRIPTION
**Pull request by gchanan**
This also unifies the device usage between the async and sync case.

Fixes https://github.com/pytorch/pytorch/issues/10832.

### CODE REVIEW

File: test/run_test.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
     'cpp_extensions',
     'c10d',
     'cuda',
+    'cuda_primary_ctx.py'
```
Comment (SsnL) [inline test/run_test.py]: I don't think you need the `.py` here.

File: test/test_cuda_primary_ctx.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        self.assertTrue(get_is_primary_context_created(1))
+
+        y = torch.randn(1, device='cpu')
+        y.copy_(x)
+
+        # We should still have only created context on 'cuda:1'
+        self.assertFalse(get_is_primary_context_created(0))
+        self.assertTrue(get_is_primary_context_created(1))
+
+    # DO NOT ADD ANY OTHER TESTS HERE!  ABOVE TEST REQUIRES FRESH PROCESS
```
Comment (SsnL) [inline test/test_cuda_primary_ctx.py]: Not if we do `cuDevicePrimaryCtxReset` at `setUp`. :) 

Certainly I hope this test covers more cases. Now it only checks if we mess up in copy (and print I guess). But this bug can happen elsewhere too.

Comment (SsnL) [reply test/test_cuda_primary_ctx.py]: Btw, does folding into `test_cuda` not work?

Comment (gchanan) [reply test/test_cuda_primary_ctx.py]: I don't think you can guarantee the test order there unless you shell out a new command, but that seems overly complicated.

File: aten/src/THC/generic/THCTensorCopy.cpp [hunk 3]
```diff
@@ -58,6 +58,13 @@ void THTensor_(copyCuda)(THCState *state, THTensor *self, struct THCTensor *src)
 
   {
     THTensor *selfc = THTensor_(newContiguous)(self);
+    int tensorDevice = THCTensor_(getDevice)(state, src);
```
Comment (ezyang) [inline aten/src/THC/generic/THCTensorCopy.cpp]: Here would be a good place to put a note about how to do this correctly re current device.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**